### DISCUSSION TITLE
Customize Installer

### DISCUSSION THREAD

**Discussion by mcbglug** [mickem/nscp#826] [Q&A]
Is there a way to add some scripts, and modify nsclient.ini in the MSI installer? rebuilding the MSI or in some other ways?

**Comment by mickem** [comment]
Sorry for the late reply. You do not need to rebuild the MSI for either of those; the installer has public properties for exactly this, and the settings mechanism can pull in extra files.

**Modifying `nsclient.ini` at install time**

- `CONF_SET=section;key;value;section;key;value...` writes individual settings, for example:

  ```
  msiexec /qn /i NSCP-<version>-x64.msi CONF_SET="/settings/NRPE/server;allow arguments;true;/settings/external scripts/alias;alias_cpu;check_cpu"
  ```

- `IMPORT_CONFIG=<file or URL>` copies a complete `nsclient.ini` you prepared and uses it as the configuration:

  ```
  msiexec /qn /i NSCP-<version>-x64.msi IMPORT_CONFIG=\\fileserver\deploy\nsclient.ini
  ```

- `CONF_INCLUDES=alias;file;alias;file` adds `[/includes]` entries, so the shipped `nsclient.ini` stays untouched and your additions live in a separate file that you drop next to it (or `CONFIGURATION_TYPE=http://server/nsclient.ini` if you want the agent to fetch its configuration from a central server on every start).
- `ALLOWED_HOSTS`, `NSCLIENT_PWD`, `CONF_NRPE`, `CONF_WEB`, the TLS certificate properties and more are separate properties for the common values. The full list is in the docs under [Installing NSClient++ / MSI options](https://docs.nsclient.org/setup/installing/#msi-options).

**Adding scripts**

The installer does not have a property for copying arbitrary scripts, so the usual pattern is to copy them into `scripts\custom` (or any folder) with the same deployment tooling that runs `msiexec`, and register them either from an included ini file or with `CONF_SET`:

```
CONF_SET="/settings/external scripts/scripts;check_foo;scripts\custom\check_foo.bat"
```

The `[/settings/external scripts]` section also accepts `script path = scripts\custom` to load every script in a folder, which saves listing them one by one.

**If you really want one self-contained MSI**

Two options: build a transform (`.mst`) with Orca or a similar MSI editor that sets the properties above and adds your files, and install with `msiexec /i NSCP.msi TRANSFORMS=custom.mst`; or rebuild from source, the WiX project is under `installers/installer-NSCP` in this repository. Both are more work to maintain than the property-based approach, which survives upgrades unchanged.
